Topics/jupyter-notebook.md

```diff
@@ -0,0 +1,154 @@
+
+An alternative possibility is to use [**Google Colaboratory**](https://colab.research.google.com/notebooks/welcome.ipynb).
+
+Colaboratory is a free Jupyter notebook environment that requires no setup and runs entirely in the cloud. 
+
+With Colaboratory you can write and execute code, save and share your analyses, and access powerful computing resources, all for free from your browser. 
+
+### Considerations
+
+As we said from the beginning, Jupyter Notebooks are great for exploring data, doing analysis and building small prototypes, so you can quickly try many different approaches to the same problem.
+
+But beware! 
+
+You can't think that the final working system you built (presumably a Machine Learning model that makes predictions) is based entirely on a Jupyter Notebook!
+
+_In fact, putting models into production in the real world with Notebooks is a totally wrong practice_.
+
+There are even more extreme opinions, i.e. those who do not recommend their use at all, like Joel Grus in [**I don't like notebooks.**](https://www.youtube.com/watch?v=7jiPeIFXb6U)
+
+See also: [Pitfalls of Jupyter Notebooks](https://scicomp.aalto.fi/scicomp/jupyter-pitfalls.html), it can save you a lot of headache!
+
+We at Virgilio think that _they are not the extreme evil_ as Grus suggests, but they certainly do not encourage the principles of software engineering or code cleaning in general.
+
+That's why **it's crucial to know when to stop using the Notebook and start refactorizing the code into modular Python scripts**, which can then be conveniently used for purposes beyond exploration and experimentation.
+
+Check the [nbconvert](https://nbconvert.readthedocs.io/en/latest/) module, and see also: [How do I convert a IPython Notebook into a Python file via command line?](https://stackoverflow.com/questions/17077494/how-do-i-convert-a-ipython-notebook-into-a-python-file-via-commandline)
+
+Check the [nbdime](https://github.com/jupyter/nbdime), it provides tools for diffing and merging of Jupyter Notebooks.
+
+### Tips and Tricks
+
+Like any other new tool you learn, there are often little tricks or conveniences hidden in the documentation that can really make life easier if you use the tool in question every day!
+
+The Jupyter Notebooks are no less, so here's to you a very detailed article on the tips and tricks of Jupyter!
+
+See [**Jupyter Notebook Tips, Tricks, and Shortcuts**](https://www.dataquest.io/blog/jupyter-notebook-tips-tricks-shortcuts/)
+
+### Notebooks Examples
+
+Spread across the Web you can probably find hundreds of thousands of Jupyter Notebooks, but how do you navigate them all in an organized way?
+
+Luckily someone did this job before Virgil, and created this incredible resource:
+
+[**A gallery of interesting Jupyter Notebooks**](https://github.com/jupyter/jupyter/wiki/A-gallery-of-interesting-Jupyter-Notebooks)
+
+Take your time and explore this rich repository, you will surely find something that interests you! 
+
+### Conclusions
```

Comment:
I suggest you share other forms of Jupyter notebooks via these links:
- [JuPyteR example running inside a docker container - able to run Java code as well](https://github.com/neomatrix369/awesome-ai-ml-dl/tree/master/examples/JuPyteR)
- [Apache Zeppelin running inside a docker container - able to run Scala code](https://github.com/neomatrix369/awesome-ai-ml-dl/tree/master/examples/apache-zeppelin)

Both the above can run many other languages inside the notebook if we install the right extension or interpreter or plugin for them. 

Also see [additional sources of kernels](https://github.com/neomatrix369/awesome-ai-ml-dl/blob/master/examples/JuPyteR/README.md#additional-source-of-kernels).

Try to use these links creatively, its good to give readers an idea of what else is there in this space as there are misconceptions about tools and their usages.

Reply:
I think these topics are too advanced for now. I'm starting thinking about a guide on Containers and Docker though :) because it's very important, but a beginner doesn't need it imo...